### ISSUE TITLE
REST controllers with JSON output

### ISSUE THREAD

**Issue by AnandMuley** [psi-probe/psi-probe#470]
Hi,

I checked the probe code. I could not find any REST controllers which will give me probe output as a JSON or XML output for using the data in other UI plotting application. Would you please confirm if there is any such provision in probe or not.

**Comment by MALfunction84** [comment]
There is not, although discussion in #456 brought this up as a possibility in 3.0.

**Comment by AnandMuley** [comment]
Thank You for the update. Another query on the similar lines is that, does probe register/expose some Managed Beans so that I can fiddle with that for fetching the data from probe using jconsole, etc.using JMX.

**Comment by MALfunction84** [comment]
@AnandMuley Not yet. That's #13.

**Comment by AnandMuley** [comment]
Ok...Thank You !

**Comment by MALfunction84** [comment]
This hasn't been implemented yet. Reopening.

**Comment by kevkon3** [comment]
This might be something I would be interesting in working on. I assume the request is to make a Web API for accessing stuff that the probe web page shows? Would we be able to come up with a list of features desired for the Web API? 

More importantly, do we see this as something that would actually be used?

**Comment by jander99** [comment]
Hi Kevin.

Let's move this to the discussion forums.  I will start a topic and link it
back here and we can star laying out an API design.  Ideally I'm sure the
UI components could utilize the API.

**Comment by jander99** [comment]
https://groups.google.com/forum/#!topic/psi-probe-discuss/gwv9Bh4fTYA

**Comment by diogosantana** [comment]
Is this still on the radar for 3.0.0 since we at M2?

**Comment by hazendaz** [comment]
I think we may wait until tomcat 9 is official before 3 is released.  Milestones are more so I can get a somewhat official into a production system.  So this certainly is still on the plate for version 3.  I'd say everything is.


Sent from Outlook Mobile






On Sun, Feb 7, 2016 at 1:50 AM -0800, "diogosantana" <notifications@github.com> wrote:





Is this still on the radar for 3.0.0 since we at M2?

---

**Comment by diogosantana** [comment]
Ok. So, how we will implement this? Spring support for rest or another more specialized API like CXF?

**Comment by jander99** [comment]
My vote is Spring MVC.  Eventually I would like to transition from using
ModelAndView to having REST endpoints that can feed a UI, then start
swapping out parts of the UI.  We're currently looking at using Slack for
some collaboration, and may be releasing some news in the readme or on the
.io site in the coming weeks.

**Comment by diogosantana** [comment]
Which .io site?

**Comment by jander99** [comment]
http://psi-probe.github.io/  eventually. :)

**Comment by jander99** [comment]
@hazendaz @MALfunction84 what do you think? New module or include in web?

**Comment by hazendaz** [comment]
Is the idea that we provide output via REST as JSON and per page the thought is listing what we have available?  Personally, I'd use Jersey/Moxy or Jackson to pull this off.  A new module might make sense so there is less to scan at startup for REST services.  Then they just call into various points as do the JSP pages then producing JSON.  Ultimately we would use these too and replace the entire front end so there is no more JSP and use something like angular and/or bootstrap.

**Comment by jander99** [comment]
Yes my thought is that we can write REST endpoints to all the data currently being presented in the JSPs, then slowly rework the presentation layer to use them.  This will be the start of a long journey where the core logic becomes a deployable jar and the UI sits on top with the ability to connect to more than 1 set of Probe-enabled containers.

**Comment by hazendaz** [comment]
Perfect!  We are on the same page.  My vote goes to separate module.

**Comment by MALfunction84** [comment]
This all sounds good to me.

To start with, we should really consider condensing all of PSI Probe's functionality down into a series of interfaces which transact using primitives and POJOs. This would be a true public API for PSI Probe. Primitives will serialize cleanly to JSON, and we can inject such an interface into a thin REST service implementation. This would also allow for straightforward endpoint implementations over JMX, JSON-RPC, or whatever the flavor of the week happens to be. (Heck, even command line.)

Additionally, the underlying implementation of the PSI Probe interface would of course start as Tomcat-specific but could expand later to include other servers, e.g. JBoss 8.0+.

**Comment by jander99** [comment]
@hazendaz @MALfunction84 slack much?

**Comment by hazendaz** [comment]
I will ...otherwise just watching git channel.

**Comment by jander99** [comment]
Looks like the filter chain is getting in the way of the rest paths.  If anyone is more familiar with how Spring security and Sitemesh work, please check out my attempt at https://github.com/jander99/psi-probe/tree/rest

**Comment by MALfunction84** [comment]
You can probably drop Sitemesh entirely.

**Comment by jander99** [comment]
Then how will the existing UI work while we make all the necessary changes
underneath to turn this into a decoupled webapp?

**Comment by MALfunction84** [comment]
Oh, it wouldn't. I wasn't aware you were trying to have both work at once.

You could try replacing the current url-pattern for sitemesh to something like this:

    <url-pattern>/</url-pattern>
    <url-pattern>*.htm</url-pattern>
    <url-pattern>*.ajax</url-pattern>

**Comment by hazendaz** [comment]
Due to no activity on this, I have switched it to backlogged.  It will be significantly easier later once we do a number of other configuration changes so this can be a 3.1 or later addition.  It's definitely wanted as it will allow psi probe to be used in a spring boot fashion but without any work and release soon for 3.0, it's best to shelve for time being.

**Comment by rockneverdies55** [comment]
Any updates on the rest api addition ?

**Comment by hazendaz** [comment]
no one ever provided code for it beyond the model that exists in the code base currently for spring-hateos.  if you have time feel free to look into what it would take and submit a pull request for same.  thanks.
